gcc/rust/expand/rust-expand-visitor.cc

```diff
@@ -230,37 +287,85 @@ ExpandVisitor::expand_inner_items (
+  AST::BlockExpr &expr, std::vector<std::unique_ptr<AST::Stmt>> &stmts)
 {
   expander.push_context (MacroExpander::ContextType::BLOCK);
 
   for (auto it = stmts.begin (); it != stmts.end (); it++)
     {
-      // TODO: Eventually we need to derive here as well
-
-      // auto &stmt = *it;
-
-      // if (stmt->has_outer_attrs ())
-      // {
-      //   auto traits_to_derive
-      //     = get_traits_to_derive (stmt->get_outer_attrs ());
-
-      //   // FIXME: This needs to be reworked absolutely
-      //   static const std::set<std::string> builtin_derives
-      //     = {"Clone", "Copy", "Eq", "PartialEq", "Ord", "PartialOrd"};
-
-      //   for (auto &to_derive : traits_to_derive)
-      //     if (builtin_derives.find (to_derive) != builtin_derives.end ())
-      //       {
-      // 	auto new_item = derive_item (
-      // 	  item, item->get_outer_attrs ()[0] /* FIXME: This is wrong */,
-      // 	  to_derive);
-      // 	// this inserts the derive *before* the item - is it a problem?
-      // 	it = items.insert (it, std::move (new_item));
-      //       }
-      // }
+      auto &stmt = *it;
+
+      // skip all non-item statements
+      if (stmt->get_stmt_kind () != AST::Stmt::Kind::Item)
+	continue;
+
+      auto &item = static_cast<AST::Item &> (*stmt.get ());
+
+      if (item.has_outer_attrs ())
+	{
+	  auto &attrs = item.get_outer_attrs ();
+
+	  for (auto attr_it = attrs.begin (); attr_it != attrs.end ();
+	       /* erase => No increment*/)
+	    {
+	      auto current = *attr_it;
+
+	      if (is_derive (current))
+		{
+		  attr_it = attrs.erase (attr_it);
+		  // Get traits to derive in the current attribute
+		  auto traits_to_derive = get_traits_to_derive (current);
+		  for (auto &to_derive : traits_to_derive)
+		    {
+		      auto maybe_builtin
+			= MacroBuiltin::builtins.lookup (to_derive);
+		      if (MacroBuiltin::builtins.is_iter_ok (maybe_builtin))
+			{
+			  auto new_item
+			    = builtin_derive_item (item, current,
+						   maybe_builtin->second);
+			  // this inserts the derive *before* the item - is it a
+			  // problem?
+			  it = stmts.insert (it, std::move (new_item));
+			}
+		      else
+			{
+			  auto new_items
+			    = derive_item (item, to_derive, expander);
+			  std::move (new_items.begin (), new_items.end (),
+				     std::inserter (stmts, it));
+			}
+		    }
+		}
+	      else /* Attribute */
+		{
+		  if (is_builtin (current))
+		    {
+		      attr_it++;
+		    }
+		  else
+		    {
+		      attr_it = attrs.erase (attr_it);
+		      auto new_items
+			= expand_stmt_attribute (item, current.get_path (),
+						 expander);
+		      it = stmts.erase (it);
+		      std::move (new_items.begin (), new_items.end (),
+				 std::inserter (stmts, it));
+		      // TODO: Improve this ?
+		      // item is invalid since it refers to now deleted,
+		      // cancel the loop increment and break.
+		      it--;
+		      break;
+		    }
+		}
+	    }
+	}
```

Comment:
so this is more of an observation than me requesting a change - is that code very similar to item proc macro expansion, and if yes, do you think we could factor them into one templated function? or are there some big differences I'm not seeing?

Reply:
Yes it is indeed similar, and I did consider to factor those two functions together multiple times. But there are some small differences in multiple places that make it not so easy to refactor. I'm not yet convinced the result of such refactor would make the code easier and cleaner, quite the opposite in fact. I could experiment and compare both versions in a later PR, I do not want to make this PR bigger/harder than required.